include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,245 @@
+
+    auto __fill_event = __queue.submit([&](sycl::handler& __hdl) {
+        __hdl.parallel_for<class scan_kt_init>(sycl::range<1>{__status_flags_size}, [=](const sycl::item<1>& __item) {
+            auto __id = __item.get_linear_id();
+            __status_flags[__id] = __id < __status_flag_padding ? _FlagType::__oob_value : 0;
+        });
+    });
+
+    sycl::event __prev_event = __fill_event;
+    for (int __chunk = 0; __chunk < __num_chunks; ++__chunk)
+    {
+        ::std::size_t __current_chunk_size = __chunk == __num_chunks - 1 ? __n % __chunk_size : __chunk_size;
+        ::std::size_t __current_num_wgs =
+            oneapi::dpl::__internal::__dpl_ceiling_div(__current_chunk_size, __elems_in_tile);
+        ::std::size_t __current_num_items = __current_num_wgs * __wgsize;
+
+        auto __event = __queue.submit([&](sycl::handler& __hdl) {
+            auto __tile_id_lacc = sycl::local_accessor<std::uint32_t, 1>(sycl::range<1>{1}, __hdl);
+            __hdl.depends_on(__prev_event);
+
+            oneapi::dpl::__ranges::__require_access(__hdl, __in_rng, __out_rng);
+            __hdl.parallel_for<class scan_kt_main>(
+                sycl::nd_range<1>(__current_num_items, __wgsize),
+                [=](const sycl::nd_item<1>& __item) [[intel::reqd_sub_group_size(SUBGROUP_SIZE)]] {
+                    auto __group = __item.get_group();
+                    auto __subgroup = __item.get_sub_group();
+
+                    // Obtain unique ID for this work-group that will be used in decoupled lookback
+                    if (__group.leader())
+                    {
+                        sycl::atomic_ref<_FlagStorageType, sycl::memory_order::relaxed, sycl::memory_scope::device,
+                                         sycl::access::address_space::global_space>
+                            __idx_atomic(__status_flags[__status_flags_size - 1]);
+                        __tile_id_lacc[0] = __idx_atomic.fetch_add(1);
+                    }
+                    sycl::group_barrier(__group);
+                    ::std::uint32_t __tile_id = __tile_id_lacc[0];
+
+                    // TODO: only need the cast if size is greater than 2>30, maybe specialize?
+                    ::std::size_t __current_offset = static_cast<::std::size_t>(__tile_id) * __elems_in_tile;
+                    ::std::size_t __next_offset = ((static_cast<::std::size_t>(__tile_id) + 1) * __elems_in_tile);
+                    if (__next_offset > __n)
+                        __next_offset = __n;
+                    auto __in_begin = __in_rng.begin() + __current_offset;
+                    auto __in_end = __in_rng.begin() + __next_offset;
+                    auto __out_begin = __out_rng.begin() + __current_offset;
+
+                    if (__current_offset >= __n)
+                        return;
+
+                    _Type __local_sum = sycl::joint_reduce(__group, __in_begin, __in_end, __binary_op);
```

Comment:
This limits `__binary_op` to the SYCL supported binary operators (https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:function-objects), right? The same goes for `sycl::joint_inclusive_scan()` below. Should we document this limitation?
If we want to support other operators in the future, we might want to reuse our existing reduce and scan implementations for those.

Reply:
Yes, this limitation should be documented. In the future, we can pull in the scan/reduce support for arbitrary binary operators as you suggest.